### ISSUE TITLE
Help with registering a new type of a resource (DEV question)

### ISSUE THREAD

**Issue by blurymind** [4ian/GDevelop#569]
## Description
I need to figure out how to allow gdevelop to add other filetypes than png as a resource to any project.

This is needed for two things:
- One is for pskl files - so we can preserve layer data of animations automatically when using piskel to create/edit them when editing images with piskel - also store a pskl file
- Another is for Yarn story files - In which case I want to be able to store these text files as a project resource that the newIde can create and edit (via Yarn) through a specialized yarn story game object or behavior 

The part that is confusing to me is- can one register the new resource file type with the newIDE or does it also need to be registered with the runtime (gdevelop.js) ?
 
What are the files in src that I need to be looking at in order to make this work

Construct2 for example allows to add json files as resources and later on read from them. In gdevelop - it looks like you have to manually type the filepath of any structured file as a string in order to load it via the event sheet

My plan is to first try to get this working for piskel files - and if succesful - use the knowledge to apply it to yarn files (a more complex task). So before working on Yarn story parsing - improve the current piskel workflow

**Comment by 4ian** [comment]
Resources are part of GDCore. 
Resource: https://github.com/4ian/GD/blob/master/Core/GDCore/Project/ResourcesManager.h#L33
ImageResource: https://github.com/4ian/GD/blob/master/Core/GDCore/Project/ResourcesManager.h#L165
AudioResource: https://github.com/4ian/GD/blob/master/Core/GDCore/Project/ResourcesManager.h#L236

I'm unsure if adding specific resources for Piskel/Yarn is the good way to go. I would prefer to see if we can add support for "FileResource" or "JsonResource" or anything like this.

> In gdevelop - it looks like you have to manually type the filepath of any structured file

"Files" in storage actions are a term coming from the "past", i.e: the XML files that were used in GD4 for native games. In GDJS, these "files" are stored as JSON in browser local storage.

To dig a bit more, what is exactly stored in Piskel file? Is that a JSON file?

**Comment by blurymind** [comment]
@4ian FileResource should be sufficient for both of these things imo. All I really need to keep a track of is really file paths- regardless of their extension.

In query to your question about piskel- its just a json file- so is the yarn file. They are both ordinary json text files that have specific structures.

All I really want to do here is simplify workflow in the newIde - alleviating the user from the need to manually type  paths in the event sheet in order to set up yarn data. Also allowing the ide to create/edit it easily in aid of the project without leaving gdevelop. The idea is to similarly to piskel -create a yarn document from within gdevelop and edit it inside gdevelop (via yarn).

For piskel - I need a way to associate pskl files with animations, so when the user saves an animation or a resource in piskel, the layer data is preserved in a file that is kept somewhere within the project folder

**Comment by blurymind** [comment]
Would be quite nice if we could add json files as resources in the editor. Then simply select them from a searchable drop down list in the event sheet - where applicable.

The "FileResource" type in the newIde could also get us closer to https://github.com/4ian/GD/issues/454
- for attaching .js files to gdevelop projects

**Comment by 4ian** [comment]
I've created a card here: https://trello.com/c/dT5esfRc/122-add-supporting-for-embedding-json-files-in-a-games-and-reading-them-from-events to track this feature.

JSON files would be useful if they can be read from events (useful for quiz games, I got another user asking me about this - or translations for games supporting mulitple languages).

**Comment by blurymind** [comment]
Another addon I want to make that uses json is for this library:
http://depts.washington.edu/madlab/proj/dollar/pdollar.html
Json is required to store pre-recorded gestures. 
The goal is to some day recreate this game mechanic in gdevelop:
https://www.google.com/doodles/halloween-2016
:} 
I have made this as an addon for godot btw, but it's implementation is in 1$ not p$

I exposed some of the remaining yarn data in bondage.js now, which will come in useful later for check tag/title conditions when its used as an addon
https://github.com/InfiniteAmmoInc/Yarn/pull/80

**Comment by 4ian** [comment]
I'm not sure what's your usecase of JSON here? Could you explain how they will be useful, how they will be used and how the user of GDevelop would use them with your addon?

**Comment by blurymind** [comment]
In the case of a gesture extension - for the capturing of gestures and storing them. A gesture is a point cloud that forms a line

In this case I guess that the only reason is make the extension easier to use. I could technically just have the point data in the js file.

It makes more sense for the Yarn dialogue addon that would use bondage.js - where the json file contains all the game dialogue

**Comment by zatsme** [comment]
$p support would be amazing for gdevelop, for mobile games gesture support is essential 👍🏻

**Comment by zatsme** [comment]
Yes, but using the library will be easier and probably more accurate. 😁

**Comment by 4ian** [comment]
Do you have an idea of how to expose the library to the users in a way that would make it easy to use? :) Like, should there be some kind of interface to author the gestures, or even a list of predefined gestures?

**Comment by blurymind** [comment]
@4ian in the case of how I made it in godot - I included a gesture recorder with the addon. This can be done i gdevelop in the form of the example project file. The gesture recorder can save/load strokes from/to a json file- which the addon can be set to load :)

**Comment by 4ian** [comment]
I see! It's an interesting and really good idea to have the gesture recorder as part of the example.
(btw, your addon for godot looks nice, good job! :))

I think it's still possible to start with a set of predefined set of gestures and by then I'll maybe have the time to see if we can include external JSON files.

It seems that your gesture recorder would also need to save local files - which is not possible for now. So this is a different addition to do too (i.e: dealing with local files for storage) that would work only for games exported on Windows/mac/Linux (i.e: with Electron). Not sure exactly what's the best way to integrate this to the existing storage actions. 🤔

**Comment by blurymind** [comment]
Ah thank you :) 

can we not save game states as json files on android?
Saving to mobile is something to tackle with image resources too 

An alternative solution is to display the result as a long string that the developer can copy - if it fails to save

In any case, a custom gesture editor would be nice, and shouldnt be too hard to implement - even as a web version. But its not as important as being able to use json files as a resource on any platform.

This affects saving piskel layers and dealing with dialogue data - which are more important to tackle atm imo

**Comment by blurymind** [comment]
Btw if we use piskel's json data to store image data, instead of actual bitmaps we use now- we can technically store it in the json project file. That way when the user downloads their project file from the web version - it will contain all of their animations with layers and everything!

Downside to this is that it will theoretically increase the size of project files. You are also trying to move away from having the entire project on a single file

**Comment by blurymind** [comment]
For piskel - I just need a way to save the piskel object - either to an external json file- stored in the project folder, or as part of the actual project resource json file. It's only goal is to preserve the layer data of the bitmaps piskel created.
In theory we can even stringify any imported images that are not made in piskel and store them in the project json. But my goal really is to solve the layers problem atm.

For Yarn -  I have the yarn editor- which is very good for authoring non linear dialogue and even setting up cut scene events while the dialogue is running. It saves the same way as piskel - I can either send the whole dialogue back to gdevelop's project json or to a json file resource. In this case I think that it has to be outside of the project json - as it will make it much easier to deal with when translating a game or collaborating on a story with non programmers.
There is also the runtime part of it- for which I want to use bondage.js to parse the data via a js extension. The js extension will require to access json resource files that were set  to be used in the newIde. That is the biggest roadblock atm. From there on actually parsing them should be easy.
Bondage.js now gives access to everything I need to set up the actions/conditions to be used in gdevelop, but I must first initiate it by giving it the yarn story object

**Comment by Lizard-13** [comment]
Not sure if it was suggested already for this very specific feature, but if it just needs to store the path points you could save the points in the project file as any object data, this way you can let the user draw the gestures in the object editor, get the points (saved in the project with the object data) and load it directly with the data passed to the runtime object. I say it because unlike piskel and yarn it doesn't even need external files, because there is no editor for it at all.

**Comment by blurymind** [comment]
@Lizard-13 That is a cool idea that I want to try with piskel objects, but I have no idea how to write new custom variables to the project.json file. Can you show me any examples? :}

**Comment by Lizard-13** [comment]
In the old-extensions system it was made adding attributes to the object un/serializer functions: https://github.com/4ian/GD/blob/8cad5d76463eb059994d5bf68a726b6f4826c9a6/Extensions/TiledSpriteObject/TiledSpriteObject.cpp#L38-L50

And then the data is passed to the runtime object:
https://github.com/4ian/GD/blob/8cad5d76463eb059994d5bf68a726b6f4826c9a6/Extensions/TiledSpriteObject/tiledspriteruntimeobject.js#L19-L25

But no idea how it is done in the new pure js extensions, surely @4ian knows if/how it can be down :)

**Comment by 4ian** [comment]
What @Lizard-13  mentioned works but it's for objects or behaviors only :)

@blurymind For Piskel:

I have being thinking about it, and one thing to consider is that resources are resources **for the game**, not for the editor. So I think that for Piskel, using "JSON" resources is not appropriate: we don't want to clutter the final games about information about piskel layers. Instead, maybe we could add to `gd.ImageResource` some kind of arbitrary extra "data" that could be attached to it. This would allow you to have a piskel property where you can store the json of the layers.
Would that work for you?
As a reminder, `gd.ImageResource` is the class used to store images in the list of resources. It's used there when Piskel is closed:

https://github.com/4ian/GD/blob/6cdee3b9773045ad2daa645a8c47aa71878fe83c/newIDE/app/src/ResourcesList/LocalPiskelBridge.js#L48-L61

If we image that we have something like `imageResource.setMetadata("piskel", thejsonofpiskel)`, would that work? (You would have to load it to when opening Piskel).

For yarn:
Is having an object something that is possible? In this case, the JSON of the dialog would be stored inside the object.
If that's not the case, let's image we have a JSON resource:
* When would you save them? What would be the name of the resource?
* In the game engine, how would you load it? 

For the last question, here is how the gdjs.ImageManager and gdjs.SoundManager are constructed:
https://github.com/4ian/GD/blob/e9bf8ff30963d25b107c45e5e527c8f6a869068d/GDJS/Runtime/runtimegame.js#L20-L25

They take a reference to the resources and then iterate over them to load (https://github.com/4ian/GD/blob/e9bf8ff30963d25b107c45e5e527c8f6a869068d/GDJS/Runtime/pixi-renderers/pixi-image-manager.js).

Do you have an idea of what kind of API/function/method you'd like to have?

**Comment by blurymind** [comment]
@4ian this is a great idea!
For the piskel you are right. It would make alot of sense to store the data in the actual imageresource object - as metadata of the image sequence. If you add the ability to set metadata like that- I will put it in use when piskel loads/saves in gd5 if you want to :) 👍 

For yarn - as a resource having them would be very helpful- as it will enable people who only have yarn to open the files, edit them and send them over to the gdevelop dev. It will make story translation much easier.
My vision for this is to have Yarn editor itself embeded with gdevelop - very much like piskel and monaco. In yarn's case, it will treat json files the same way piskel treats image sequences- as resources. If there is no yarn (json) file loaded- open yarn and when saving-create one. The name of the file can correspond to the name of the game object it is being attached to, but can also be specified by the user.
After saved (to project folder) and attached to a game object - to edit it, the user doesnt need to navigate to it.Its a single click open and single click save and close. 
On the runtime's side- the json file gets loaded into a dictionary{} when the game loads. It is initiated by the dev via the event sheet. The dev can setup in the event sheet and yarn additional triggers while a dialogue is running. For example, tags that were set in yarn can be used as conditions in the event sheet. My implementation also listens for when a text card has changed- what choices were made by the player, what hidden tags were set in that dialogue card and even commands.

Basically I want to load the json files that are attached as a resource - as an ordinary object{} on init, so my extension wrapper of bondage.js can use its data to do things. Json files are tiny, even if you have a lot of data in them. The concept is the same as how image resource and sound resource is loaded and used when triggered from whats in the event sheet :)

**Comment by KinkGD** [comment]
@blurymind , any news from "the front" ?

**Comment by blurymind** [comment]
I will revisit this once @4ian adds support for metadata in objects like this
`imageResource.setMetadata("piskel", thejsonofpiskel)`
When we have that method merged, I will address saving piskel data and storing the layers :)

**Comment by 4ian** [comment]
Support added :)
This interface is `resource.setMetadata(string)` but for now please store JSON inside, inside an object containing the name of the editor. This way we could later change the interface to be `resource.setMetadata(editorName, string)`

**Comment by blurymind** [comment]
@4ian thank you! 😄

**Comment by blurymind** [comment]
I am currently refactoring the external editor handling code to check if metadata exists and if so, to pass it accordingly to the editor

**Comment by 4ian** [comment]
Do you think we can close this? I think metadata was the proper way of solving this :)

**Comment by 4ian** [comment]
Let's talk about it back when you want to work on this if that's ok?
Might be metadata or some other way to store things yeah.

**Comment by blurymind** [comment]
Yes ok 😄 Closing this now
